feat: Filter LabelValues calls with matchers

jdbaldry opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

jdbaldry committed:
feat: Filter LabelValues calls with matchers

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
test: Add test for filteredMatchers being sent to downstream queriers

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jbvmio committed:
Use non APIPA when assigning IP Address (#4032)

* use APIPA as last resort

Signed-off-by: Jimmy Bonds <jbonds@jbvm.io>

* Update CHANGELOG.md

Signed-off-by: Jimmy Bonds <jbonds@jbvm.io>

* tidy up changelog

Signed-off-by: Jimmy Bonds <jbonds@jbvm.io>

* avoid empty value overwrites, preserving APIPA

Signed-off-by: Jimmy Bonds <jbonds@jbvm.io>

* remove duplicates

Signed-off-by: Jimmy Bonds <jbonds@jbvm.io>

stevesg committed:
Add new metrics to aid troubleshooting tombstone convergence. (#4231)

* Add new metrics to aid troubleshooting tombstone convergence.

- `memberlist_client_kv_store_value_tombstones`
    Expose the number of tombstones currently in the value for each key
    in the store. If the tombstones are being propagated as expected,
    then each instance should have the same value for this metric.
- `memberlist_client_kv_store_value_tombstones_removed_total`
    Count the number of tombstones which have been removed because they
    aged out. This can be used to troubleshoot whether tombstones are
    being removed too soon (i.e. before the state is converged).
- `memberlist_client_messages_to_broadcast_dropped_total`
    Count occurrences of messages being quietly dropped when they were
    expecting to have been broadcast. As log messages already existed
    for these events, the logs have had the `key` field added instead
    of exposing per-key metrics, for consistency with other metrics
    of this nature in the package.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

* Review comments.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

simonswine commented on pkg/querier/tenantfederation/merge_queryable.go:
```diff
@@ -139,8 +139,19 @@ type mergeQuerier struct {
 	log, _ := spanlogger.New(m.ctx, "mergeQuerier.LabelValues")
 	defer log.Span.Finish()
+
+	matchedValues, filteredMatchers := filterValuesByMatchers(m.idLabelName, m.ids, matchers...)
+	var labelValues []string
+	var ids []int
+	for labelPos := range m.ids {
```
I think it would be way easier to pass the map `matchedValues` to a method `mergeDistinctStringSliceWithLabelValues` (with a nil map returning everything).

simonswine commented on pkg/querier/tenantfederation/merge_queryable.go:
```diff
@@ -139,8 +139,19 @@ type mergeQuerier struct {
+
 	if name == m.idLabelName {
-		return m.ids, nil, nil
+		return labelValues, nil, nil
```
I suggest you build `labelValues` only if you are within in this branch

simonswine reviewed (commented):
Thank you, this is almost there.

I think the conversion of the `map[string]` to a `[]int` is avoidable and I would simply pass the map around...

simonswine commented:
I also think you can cleanup the skip logic in the tests with this change

Commit:
Replace validation error with error in query limiter (#4240)

* Replace validation error with error in query limiter

Signed-off-by: Tyler Reid <tyler.reid@grafana.com>

* Change query limiter integration test to check for 500 rather than contained string

Signed-off-by: Tyler Reid <tyler.reid@grafana.com>

jdbaldry committed:
refactor: Remove use of id int slice

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

simonswine commented on pkg/querier/tenantfederation/merge_queryable.go:
```diff
@@ -139,8 +139,17 @@ type mergeQuerier struct {
+
 	if name == m.idLabelName {
-		return m.ids, nil, nil
+		var labelValues []string
```
You could create the slice in with the correct capacity from the start
```suggestion
		var labelValues = make([]string,0, len(matchedValues))
```

simonswine commented on pkg/querier/tenantfederation/merge_queryable.go:
```diff
@@ -200,17 +209,20 @@ type stringSliceFuncJob struct {
-			querier: m.queriers[pos],
-			id:      m.ids[pos],
+	for pos, id := range m.ids {
+		if _, matched := labelValues[id]; matched || labelValues == nil {
```
I am actually surprised that this isn't a panic on a nil map

simonswine reviewed (approved):
Apart from the mentioned nits I think this is ready to go upstream

jtlisi committed:
Clarify limits flags (#4243)

jdbaldry committed:
fix: Create slice with correct capacity

Co-authored-by: Christian Simon <simon@swine.de>
Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

gouthamve committed:
nit: fix typo in metric variable name (#4249)

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

pstibrany committed:
Added target to fetch build image, instead of building it locally. (#4244)

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

56quarters committed:
Ensure missing runtime ingester settings are nil (#4246)

* Ensure missing runtime ingester settings are nil

Make sure that omitted or missing ingester configuration that is
part of runtime configuration is treated as `nil` instead of an
empty object. This fixes an issue where the settings in the empty
object were overriding global default ingester limits.

Fixes #4228

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

* Update CHANGELOG.md

Signed-off-by: Marco Pracucci <marco@pracucci.com>

Co-authored-by: Marco Pracucci <marco@pracucci.com>

pracucci committed:
Fixed regression introduced in #4240 (#4251)

Signed-off-by: Marco Pracucci <marco@pracucci.com>

pracucci committed:
Fix ruler storage single binary mode (#4252)

* Fix ruler single-binary when new ruler storage config is used

Signed-off-by: Marco Pracucci <marco@pracucci.com>

* Add CHANGELOG entry

Signed-off-by: Marco Pracucci <marco@pracucci.com>

* Update CHANGELOG.md

Signed-off-by: Marco Pracucci <marco@pracucci.com>

Co-authored-by: gotjosh <josue.abreu@gmail.com>

* Added unit test and switched getting started config to new ruler_storage config

Signed-off-by: Marco Pracucci <marco@pracucci.com>

Co-authored-by: gotjosh <josue.abreu@gmail.com>

pstibrany committed:
Update Alertmanager to latest master (f686ff3be2bf). (#4237)

* Update Alertmanager to latest master (f686ff3be2bf).

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Run go mod tidy, go mod vendor again.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Update Alertmanager to latest master.
Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Fix compilation errors after latest update.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Update prometheus/common repository to 6ef301f414bf5ea9667affde57bffb479cc2511e.
Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

pracucci committed:
Mark new querier limits as experimental (#4256)

Signed-off-by: Marco Pracucci <marco@pracucci.com>

56quarters committed:
Enable active series metrics in the ingester by default (#4257)

* Enable active series metrics in the ingester by default

This change calculates and exports the `cortex_ingester_active_series`
by default. Up to this point, the metric was disabled by default since
calculating it consumes some amount of memory.

The original PR (#3153) estimated for 1M active series at least 40MB
and up to another 200MB depending on our luck reusing labels from the
ref cache.

We (Grafana) have been running with this setting enabled on our ingesters
for some time and the resource usage doesn't appear to be significant.
This feature appears to add between 1.2 - 1.6% in memory usage when
enabled: ~140MB out of a total of ~10GB of memory used per ingester.

The ingesters I measured this on

* Have multiple tenants running production workloads
* Have about 1.3M active series each
* Have about a 10GB working set (as measured by `kubectl top` and exported
  k8s metrics)

Based on this and the utility of the metric itself, I'd like to enable it
by default.

Screenshots of the pprof heap output attached.

Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

jdbaldry committed:
test: refactor mergequeryable tests along method boundaries (#4202)

* fixup: Document TestMergeQueryable

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* test: Refactor tests along Querier methods

This refactors tests to run along the Querier.Select,
LabelQuerier.LabelNames, and LabelQuerier.LabelValues methods to make
implementing matchers in LabelValues more easily tested.

It uses "Scenarios" to consistently configure mockTenantQueryables that are
upstream to the MergeQueryable being tested across each set of tests.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* test: Add names to all LabelValues tests

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* test: Make test names more consistent

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

* test: Address review feedback

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

pracucci committed:
Fix queries failing right after scaling up store-gateways (#4263)

Signed-off-by: Marco Pracucci <marco@pracucci.com>

pstibrany committed:
Dispatcher groups limits (#4254)

* Expose dispatcher aggregation groups limit.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Expose dispatcher aggregation groups limit reached metric.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* CHANGELOG.md

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Send two alerts per alert group.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* Simplify option name, add alertmanager limits to list of experimental features.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* Fix metric name.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

pstibrany committed:
Alertmanager alerts limits (#4253)

* Add store limits.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Expose alerts limiter metrics.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Fix tests.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* CHANGELOG.md

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

* Address review feedback.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* Added comment.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* Address review feedback.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* Move check to the top.

Signed-off-by: Peter Štibraný <pstibrany@gmail.com>

* When existing alert grows and doesn't fit the size limit anymore, it is rejected.

Signed-off-by: Peter Štibraný <peter.stibrany@grafana.com>

stevesg committed:
Fix default memberlist configuration value for RetransmitMult. (#4269)

* Fix default memberlist configuration value for RetransmitMult.

If configuration is not explicitly given for RetransmitMult (via
`-memberlist.retransmit_factor`), then it is intended to be picked up
from `DefaultLANConfig`. However, though the correct value was being
used to configure `memberlist` itself, zero would be passed into the
`TransmitLimitedQueue` used for broadcasting ring updates. This
essentially means that ring updates are only ever gossiped once.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

* Add simplified integration test case.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

* Changelog.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

* Review comments.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>

* Fix to race condition in unit test.

The test was shutting down the KV store then attempting to read form it.
Sometimes this would work if the KV took some time to shutdown, which it
often will, but if it shuts down quickly, then the read will fail.

Signed-off-by: Steve Simpson <steve.simpson@grafana.com>
